verovio-util/verovio-proxy.js

```diff
@@ -4,200 +4,258 @@ verovio.vrvToolkit = verovio.vrvToolkit || {};
 
-verovio.toolkit = function () {
-	this.ptr = verovio.vrvToolkit.constructor();
-	console.debug("Creating toolkit instance");
-	verovio.instances.push(this.ptr);
+verovio.toolkit = function ()
+{
+    this.ptr = verovio.vrvToolkit.constructor();
+    console.debug( "Creating toolkit instance" );
+    verovio.instances.push( this.ptr );
 }
 
-verovio.toolkit.prototype.destroy = function () {
-	verovio.instances.splice(verovio.instances.indexOf(this.ptr), 1);
-	console.debug("Deleting toolkit instance");
-	verovio.vrvToolkit.destructor(this.ptr);
+verovio.toolkit.prototype.destroy = function ()
+{
+    verovio.instances.splice( verovio.instances.indexOf( this.ptr ), 1 );
+    console.debug( "Deleting toolkit instance" );
+    verovio.vrvToolkit.destructor( this.ptr );
+};
+
+verovio.toolkit.prototype.edit = function ( editorAction )
+{
+    return verovio.vrvToolkit.edit( this.ptr, JSON.stringify( editorAction ) );
+};
+
+verovio.toolkit.prototype.editInfo = function ()
+{
+    return JSON.parse( verovio.vrvToolkit.editInfo( this.ptr ) );
 };
 
-verovio.toolkit.prototype.edit = function (editorAction) {
-	return verovio.vrvToolkit.edit(this.ptr, JSON.stringify(editorAction));
+verovio.toolkit.prototype.getAvailableOptions = function ()
+{
+    return JSON.parse( verovio.vrvToolkit.getAvailableOptions( this.ptr ) );
 };
 
-verovio.toolkit.prototype.editInfo = function () {
-        return verovio.vrvToolkit.editInfo(this.ptr);
+verovio.toolkit.prototype.getElementAttr = function ( xmlId )
+{
+    return JSON.parse( verovio.vrvToolkit.getElementAttr( this.ptr, xmlId ) );
 };
 
-verovio.toolkit.prototype.getAvailableOptions = function () {
-	return JSON.parse(verovio.vrvToolkit.getAvailableOptions(this.ptr));
+verovio.toolkit.prototype.getElementsAtTime = function ( millisec )
+{
+    return JSON.parse( verovio.vrvToolkit.getElementsAtTime( this.ptr, millisec ) );
 };
 
-verovio.toolkit.prototype.getElementAttr = function (xmlId) {
-	return JSON.parse(verovio.vrvToolkit.getElementAttr(this.ptr, xmlId));
+verovio.toolkit.prototype.getExpansionIdsForElement = function ( xmlId )
+{
+    return JSON.parse( verovio.vrvToolkit.getExpansionIdsForElement( this.ptr, xmlId ) );
 };
 
-verovio.toolkit.prototype.getElementsAtTime = function (millisec) {
-	return JSON.parse(verovio.vrvToolkit.getElementsAtTime(this.ptr, millisec));
+verovio.toolkit.prototype.getHumdrum = function ()
+{
+    return verovio.vrvToolkit.getHumdrum( this.ptr );
 };
 
-verovio.toolkit.prototype.getHumdrum = function () {
-	return verovio.vrvToolkit.getHumdrum(this.ptr);
+verovio.toolkit.prototype.getLog = function ()
+{
+    return verovio.vrvToolkit.getLog( this.ptr );
 };
 
-verovio.toolkit.prototype.getLog = function () {
-	return verovio.vrvToolkit.getLog(this.ptr);
+verovio.toolkit.prototype.getMEI = function ( param1, scoreBased )
+{
+    if ( typeof param1 === 'undefined' )
+    {
+        return verovio.vrvToolkit.getMEI( this.ptr, JSON.stringify( {} ) );
+    }
```

Comment:
I don't understand what's happening here, if the type of ``param1`` is undefined you return an empty JSON?

Reply:
This isn't a file we manage (it's part of the compilation of verovio into both an NPM module and a web toolkit) but basically the `getMEI` function expects a string that can be parsed to a JSON object. Since this wasn't always the case and no options were once accepted, this just gives it an empty JSON object string rather than nothing to prevent an error. Why it does `JSON.stringify( {} )` instead of `"{}"`, your guess is as good as mine.

Reply:
Ok, but what does it mean to check ``typeof param1 ===  'undefined'``? Is it the same as checking for ``null``? (My lack of JS knowledge shows here)

Reply:
No problem. It's similar to checking for `null`, but not quite. So `undefined` is a value assigned to JS variables before assignment, while `null` must be explicitly defined. Since JS does dynamic typing, a undefined value doesn't have a type yet and is evaluated as having the type `undefined`. JS functions can be called without all or any of their parameters. Unless a default is given, these parameters are treated as undefined and so in this case if the function is just called `param1` is `undefined`. So this block is just checking if a value was given for `param1` and, if not, calling the backend function with an empty JSON object as a default.

Reply:
I see, thanks!